### DISCUSSION TITLE
Regarding Credentials & Certificates

### DISCUSSION THREAD

**Discussion by jmg011** [NetApp/harvest#884] [Q&A] [answered]
1) Is there a way to integrate Vault credentials in Harvest.yml configuration file instead of using clear text username and password?
2) Regarding certificates as an alternative to username/password, is there a documentation on how to set it up for both cmode and 7mode?

**Comment by rahulguptajss** [comment]
@jmg011 
1.  Does https://github.com/NetApp/harvest/wiki/Harvest-and-Secrets-Management-with-SOPs help?
2. Certificates https://netapp.github.io/harvest/latest/prepare-cdot-clusters/#using-certificate-authentication

**Reply by jmg011** [reply]
@rahulguptajss @cgrinds  Thank you. File encryption is a good idea and certificate is better. But 7modes won't support certificate as @cgrinds mentioned, so we do need to go with encryption. However, Is there any other way to connect with the Vault? Like harvest can query Vault API to get password/token?

**Comment by cgrinds** [comment, marked as answer]
hi @jmg011 there are many auth implementations especially across the managed environements that Harvest runs. Which one are you interested in?

7mode does not support client certificates so nothing Harvest can do there. The [SOP](https://github.com/NetApp/harvest/wiki/Harvest-and-Secrets-Management-with-SOPs) link Rahul pasted above was created for a few customers using 7mode that didn't want their passwords in the clear.

**Reply by cgrinds** [reply]
we'll can take a look and to be clear you're looking for a solution that works with 7-mode and uses Hashicorp's [Vault](https://github.com/hashicorp/vault). Is that correct?

**Reply by jmg011** [reply]
@cgrinds yes, that's correct.

**Reply by cgrinds** [reply]
hi @jmg011 the team meet and discussed this request after evaluating the options. Can you give this a read and let us know if this addresses your need? We'll add support for Vault via [vault agent templates](https://www.vaultproject.io/docs/agent/template). We'll include more details in the pull request, but the gist is you can setup [vault agent](https://www.vaultproject.io/docs/agent) auto-auth to automatically authenticate to Vault and manage the token renewal process for locally-retrieved dynamic/static secrets.

Harvest will include a new option to use Vault's templated credentials like this:

```
harvest.yml snippet

Pollers:
  u2:
    datacenter: dc-1
    addr: 10.193.48.11
    credentials_file: /var/secret/pollers   <==== vault created file that contains credentials
```

- Vault agent is responsible for authenticating with Vault and creating the `/var/secret/pollers` yaml file
- Harvest will read the `/var/secret/pollers` file and use the included credentials to authenticate with ONTAP. 

Here's an example of the template

```
Pollers:
    u2:
        {{- with secret "secret/harvest/u2" }}   <==== vault path to poller credentials
        username: "{{ .Data.data.username }}"
        password: "{{ .Data.data.password }}"
        {{- end }}
```

another example with ONTAP certs pulled from Vault

```
Pollers:
    jeff:
        {{- with secret "secret/harvest/jeff" }}   <==== vault path to poller certificates
        ssl_cert: {{ .Data.data.ssl_cert }}
        ssl_key:  {{ .Data.data.ssl_key }}
        {{- end }}
```

**Reply by jmg011** [reply]
@cgrinds Thank you so much for discussing this with the team and providing the possible solution. I will go through it and come back in next couple of days.

**Reply by jmg011** [reply]
@cgrinds I took some time on this. I started a thread on this and will get back soon.

**Reply by jmg011** [reply]
@cgrinds - Will the template based authentication work for default credentials or will it work for only poller level authentication?

**Reply by cgrinds** [reply]
hi @jmg011 vault agent is quite flexible since it handles the tokens, renewals, certificates, etc. With regards to default credentials - if you mean, will it work with ONTAP credentials listed in the `Defaults`section of the `harvest.yml`, the answer is yes. That's how I set it up locally. If you meant something else, can you say a bit more about what the needs are for default credentials?

**Reply by jmg011** [reply]
@cgrinds This is what I meant "_will it work with ONTAP credentials listed in the Defaults section of the harvest.yml_". Thank you for confirming. I think this approach will work for us.
